doc/developer/design/20250226_postgres_style_explain.md

```diff
@@ -0,0 +1,338 @@
+be careful to keep these tests while enabling the new behavior.
+
+[https://github.com/MaterializeInc/materialize/pull/31185
+
+## Success Criteria
+
+Our default `EXPLAIN` output should be concise and in a format
+reminiscent of Postgres's. Ideally, `EXPLAIN` output should match the
+output in `mz_lir_mapping`. The documentation should reflect this new
+syntax.
+
+## Out of Scope
+
+We are not going to build new `EXPLAIN` infrastructure, diagrams,
+etc. For example, we are not going to attempt to differentiate between
+the different meanings of `ArrangeBy` in MIR.
+
+We are not going to invent fundamentally new ways of explaining
+how Materialize works.
+
+We are not going to do a user study in advance of any changes. (But we
+will listen attentively to feedback!)
+
+## Solution Proposal
+
+Postgres explain plans have the format:
+
+```
+Operator
+  Detail
+  -> Child Operator #1
+     Detail
+     ...
+  -> Child Operator #2
+     Detail
+     ...
+```
+
+We should aim to follow Postgres's norms: operator names spelled out
+with spaces, and properties are clearly elucidated in human-readable
+formats. When it is sensible, we have simply borrowed Postgres's
+terminology, i.e., `Reduce` is renamed to `GroupAggregate`.
+
+Postgres displays some parts of the query differently from us, namely:
+
+  - Column names:
+    + When a column name is available, it just gives the name (no number).
+    + When a column name is unavailable, it gives the number using `$2`.
+  - `Map` and `Project` do not appear
+
+We will use LIR as the new default `EXPLAIN`/`EXPLAIN AS TEXT`
+output. We will update `mz_lir_mapping` to use the new Postgres-style
+syntax (fixing [a bug with `Let` and `LetRec`
+rendering](https://github.com/MaterializeInc/database-issues/issues/8993)
+in the process.
+
+We will need three pieces of work, which should all land together:
+
+ - We must implement the new output and put in appropriate SLT tests
+   for it.
+ - We must update `mz_lir_mapping` to use the new vocabulary.
+ - We must update the documentation to explain the new output, ideally
+   using this output everywhere `EXPLAIN` is used.
+
+### Concrete Mapping
+
+| LIR node    | `mz_lir_mapping` node                    | New, Postgres-style syntax              |
+| :---------- | :--------------------------------------- | :-------------------------------------- |
+| `Constant`  | `Constant`                               | `Constant`                              |
+| `Get`       | `Get::PassArrangements l0`               | `Index Scan on l0 using ...`            |
+| `Get`       | `Get::Arrangement l0 (val=...)`          | `Index Lookup on l0 using ...`          |
+| `Get`       | `Get::Collection l0`                     | `Read l0`                               |
+| `Mfp`       | `MapFilterProject`                       | `Map/Filter/Project`                    |
+| `FlatMap`   | `FlatMap`                                | `Flat Map`                              |
```

Comment:
Does Postgres have `Flat Map`? If not, we could consider `Table Function`.

Reply:
Not exactly:

```
michaelgreenberg=# EXPLAIN SELECT DISTINCT l_discount, generate_series(1, 5) from lineitem;
                               QUERY PLAN                               
------------------------------------------------------------------------
 HashAggregate  (cost=23.50..36.00 rows=1000 width=22)
   Group Key: l_discount, generate_series(1, 5)
   ->  ProjectSet  (cost=0.00..18.50 rows=1000 width=22)
         ->  Seq Scan on lineitem  (cost=0.00..12.00 rows=200 width=18)
(4 rows)
```

I'm fine with `Table Function` or `ProjectSet`.

Reply:
I vote `Table Function`!